Use smart pointers in LibWebRTCCodecsProxy.mm

rwlbuis opened the pull request:
#### e2228138170c8352130b7be99de8f971d9eb081b
<pre>
Use smart pointers in LibWebRTCCodecsProxy.mm
<a href="https://bugs.webkit.org/show_bug.cgi?id=286612">https://bugs.webkit.org/show_bug.cgi?id=286612</a>

Reviewed by Youenn Fablet.

Use smart pointers in LibWebRTCCodecsProxy.mm to address
the remaining [webkit.UncountedLambdaCapturesChecker] warnings.

* Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.mm:
(WebKit::LibWebRTCCodecsProxy::stopListeningForIPC):
(WebKit::LibWebRTCCodecsProxy::flushDecoder):
(WebKit::LibWebRTCCodecsProxy::notifyDecoderResult):
* Source/WebKit/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:

Canonical link: <a href="https://commits.webkit.org/289991@main">https://commits.webkit.org/289991@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9a3e28a63a1e6335b7955c70a52c585aaf96cbe0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88472 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7991 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42912 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/93430 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39226 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8378 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16175 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68237 "Found 4 new test failures: fast/html/details-remove-summary-3-and-click.html fast/html/details-remove-summary-4-and-click.html fast/html/details-remove-summary-5-and-click.html fast/html/details-remove-summary-6-and-click.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25958 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91474 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6416 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/80019 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48603 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6188 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38334 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76561 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/35318 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95272 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15647 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11468 "Found 1 new test failure: http/wpt/mediarecorder/MediaRecorder-matroska-AV-audio-video-dataavailable.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77102 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15903 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75875 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76362 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18829 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20754 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/19168 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/8686 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15663 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/20971 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15404 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18853 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17186 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.mm:
```diff
@@ -88,15 +88,15 @@
-        auto decoders = WTFMove(m_decoders);
+    protectedWorkQueue()->dispatch([protectedThis = WTFMove(refFromConnection)] {
+        assertIsCurrent(protectedThis->workQueue());
+        auto decoders = WTFMove(protectedThis->m_decoders);
```
I guess we should use std::exchange(protectedThis->m_decoders, { }) instead of WTFMove here and below.

youennf commented on Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.mm:
```diff
@@ -240,10 +240,10 @@ static bool validateCodecString(WebCore::VideoCodecType codecType, const String&
 void LibWebRTCCodecsProxy::flushDecoder(VideoDecoderIdentifier identifier, CompletionHandler<void()>&& completionHandler)
 {
-    doDecoderTask(identifier, [&](auto& decoder) {
+    doDecoderTask(identifier, [protectedThis = Ref { *this }, &completionHandler](auto& decoder) {
```
It should not be needed to do protectedThis since the callback is executed synchronously.
Would doDecoderTask taking a const Function<>& solve the warning?

youennf commented on Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.mm:
```diff
@@ -256,11 +256,11 @@ static bool validateCodecString(WebCore::VideoCodecType codecType, const String&
             decoder.frameRateMonitor->update();
         if (decoder.webrtcDecoder->decodeFrame(timeStamp, data)) {
-            protectedConnection()->send(Messages::LibWebRTCCodecs::FailedDecoding { identifier }, 0);
+            protectedThis->protectedConnection()->send(Messages::LibWebRTCCodecs::FailedDecoding { identifier }, 0);
```
Let's capture connection in the lambda instead.

rwlbuis replied:
Will do, I'll keep considering this option for simple lambda's going forward.

rwlbuis committed:
Use smart pointers in LibWebRTCCodecsProxy.mm
https://bugs.webkit.org/show_bug.cgi?id=286612

Reviewed by Youenn Fablet.

Use smart pointers in LibWebRTCCodecsProxy.mm to address
the remaining [webkit.UncountedLambdaCapturesChecker] warnings.

* Source/WebKit/GPUProcess/webrtc/LibWebRTCCodecsProxy.mm:
(WebKit::LibWebRTCCodecsProxy::stopListeningForIPC):
(WebKit::LibWebRTCCodecsProxy::flushDecoder):
(WebKit::LibWebRTCCodecsProxy::notifyDecoderResult):
* Source/WebKit/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:

Canonical link: https://commits.webkit.org/289991@main
